File: MetodosJavascript/Algoritmos/StrassenNaiv.js
/*
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo
*/

import { strassen_naiv_step } from "../Algoritmos/StrassenNaivStep.js";

/**
 * Realiza la multiplicación de matrices utilizando el algoritmo de Strassen de manera naiva.
 *
 * @param {Array<Array<number>>} matrizA - La primera matriz a multiplicar.
 * @param {Array<Array<number>>} matrizB - La segunda matriz a multiplicar.
 * @param {Array<Array<number>>} matrizResultado - La matriz donde se almacenará el resultado de la multiplicación.
 * @param {number} cantidadFilasMatrices - La cantidad de filas en las matrices A y Resultado.
 * @param {number} cantidadColumnasMatrices - La cantidad de columnas en las matrices B y Resultado.
 * @param {number} delimitadorMaximaIteracionesFilaColumna - La cantidad máxima de iteraciones en el bucle de fila y columna.
 * @returns {Array<Array<number>>} - La matriz resultado después de la multiplicación.
 */
export function strassen_naiv(matrizA, matrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna) {
    // Calcular el tamaño máximo para las matrices
    let tamanioMaximo = Math.max(cantidadFilasMatrices, delimitadorMaximaIteracionesFilaColumna);
    tamanioMaximo = Math.max(tamanioMaximo, cantidadFilasMatrices);

    // Ajustar el tamaño máximo para garantizar que sea potencia de 2
    if (tamanioMaximo < 16) tamanioMaximo = 16; // Si no, no es posible computar K

    // Calcular k y m
    let k = Math.floor(Math.log(tamanioMaximo) / Math.log(2)) - 4;
    let m = Math.floor(tamanioMaximo * Math.pow(2, -k)) + 1; // reducir la recursión; idea de Strassen
    let nuevoTamanio = m * Math.pow(2, k);

    // Crear nuevas matrices para almacenar los valores de matrizA, matrizB y matrizResultado
    let nuevaMatrizA = [];
    let nuevaMatrizB = [];
    let matrizResultadoAuxiliar = [];

    // Asignar memoria para cada fila de las nuevas matrices
    for (let i = 0; i < nuevoTamanio; i++) {
        nuevaMatrizA[i] = [];
        nuevaMatrizB[i] = [];
        matrizResultadoAuxiliar[i] = [];
    }

    // Inicializar cada elemento de las nuevas matrices A y B con ceros
    for (let i = 0; i < nuevoTamanio; i++) {
        for (let j = 0; j < nuevoTamanio; j++) {
            nuevaMatrizA[i][j] = 0.0;
            nuevaMatrizB[i][j] = 0.0;
        }
    }

    // Asignar valores de las matrices A y B a las nuevas matrices
    for (let i = 0; i < cantidadFilasMatrices; i++) {
        for (let j = 0; j < delimitadorMaximaIteracionesFilaColumna; j++) {
            nuevaMatrizA[i][j] = matrizA[i][j];
        }
    }

    for (let i = 0; i < delimitadorMaximaIteracionesFilaColumna; i++) {
        for (let j = 0; j < cantidadColumnasMatrices; j++) {
            nuevaMatrizB[i][j] = matrizB[i][j];
        }
    }

    // Realizar la multiplicación de matrices utilizando el algoritmo de Strassen naivo
    matrizResultadoAuxiliar = strassen_naiv_step(nuevaMatrizA, nuevaMatrizB, matrizResultadoAuxiliar, nuevoTamanio, m);

    // Copiar los resultados a la matriz resultado original
    for (let i = 0; i < cantidadFilasMatrices; i++) {
        for (let j = 0; j < cantidadColumnasMatrices; j++) {
            matrizResultado[i][j] = matrizResultadoAuxiliar[i][j];
        }
    }

    // Retorna la matriz resultado después de completar la multiplicación de matrices
    return matrizResultado;
}

File: MetodosPython/Algoritmos/StrassenNaiv.py
'''
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo

'''

from math import floor,log
import sys
sys.path.insert(0,'C:/Users/Crist/OneDrive/Escritorio/MultiplicacionMatricesGrandesProyecto/MetodosPython/Algoritmos')
"""
    Realiza la multiplicación de matrices utilizando el algoritmo de Strassen Naiv.

    Parámetros:
        matrizA (number[][]]): La primera matriz de entrada.
        matrizB (number[][]]): La segunda matriz de entrada.
        matrizResultado (number[][]]): La matriz donde se almacenará el resultado de la multiplicación.
        cantidadFilasMatrices (int): La cantidad de filas en las matrices de entrada y salida.
        cantidadColumnasMatrices (int): La cantidad de columnas en las matrices de entrada y salida.
        delimitadorMaximaIteracionesFilaColumna (int): El delimitador de la cantidad máxima de iteraciones para cada fila y columna.

    Retorna:
        (number[][]]): La matriz resultado de la multiplicación.
    """
from StrassenNaivStep import strassen_naiv_step


def strassen_naiv(matrizA, matrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna):
    # Determina el tamaño máximo necesario para las matrices de entrada

    tamanioMaximo = max(cantidadFilasMatrices, delimitadorMaximaIteracionesFilaColumna)
    tamanioMaximo = max(tamanioMaximo, cantidadFilasMatrices)
    
    # Si el tamaño máximo es menor que 16, se ajusta a 16 para garantizar la computación de K

    if tamanioMaximo < 16:
        tamanioMaximo = 16 # Si no, no es posible computar K
    
    # Calcula los parámetros para la adaptación de Strassen

    k = floor(log(tamanioMaximo)/log(2)) - 4   #Con un número igual o menor a 16 en tamaniomaximo, da 0, en caso contrario, da 1 o superior
    m = floor(tamanioMaximo * pow(2,-k)) + 1   #Con un k igual a 1 o superior, daría de una unidad mayor a tamaniomaximo para arriba
    nuevoTamanio = m * pow(2,k)
    
    nuevaMatrizA = []
    nuevaMatrizB = [] 
    matrizResultadoAuxiliar = []
    
    # Asignar memoria e inicializar Inicializar cada elemento de las nuevas matrices A y B con ceros
   

    for i in range(nuevoTamanio):
         nuevaMatrizA.append([0 for i in range(nuevoTamanio)] )
         nuevaMatrizB.append([0 for i in range(nuevoTamanio)] )
         matrizResultadoAuxiliar.append([0 for i in range(nuevoTamanio)] )
 
        
    
            
    # Asignamos en cada posición i,j de las nuevas matrices A y B los valores que están en las matrices A y B respectivamente
    for i in range(cantidadFilasMatrices):
        for j in range(delimitadorMaximaIteracionesFilaColumna):
            nuevaMatrizA[i][j] = matrizA[i][j]
            
    for i in range(delimitadorMaximaIteracionesFilaColumna):
        for j in range(cantidadColumnasMatrices):
            nuevaMatrizB[i][j] = matrizB[i][j]
    
    # Realiza la multiplicación de matrices utilizando el método Strassen

    matrizResultadoAuxiliar = strassen_naiv_step(nuevaMatrizA, nuevaMatrizB, matrizResultadoAuxiliar, nuevoTamanio, m)
    
    # Copia los resultados de la matriz auxiliar a la matriz resultado final

    for i in range(cantidadFilasMatrices):
        for j in range(cantidadColumnasMatrices):
            matrizResultado[i][j] = matrizResultadoAuxiliar[i][j]
            
    return matrizResultado #Retorna matriz resultado